### ISSUE TITLE
Performance issues with the actions scheduling on large sites

### ISSUE THREAD

**Issue by andergmartins** [publishpress/PublishPress-Future#1523]
# The problem

The problem is in the Action Scheduler async request queue runner. Here's what's happening:

1. Async request processing: When Action Scheduler processes scheduled actions, it uses admin-ajax.php requests (async requests) instead of WP-CLI or system cron.

2. The sleep() call: After processing a batch (default 25 actions), the code sleeps for 5 seconds to prevent MySQL crashes from chaining requests:

```php
protected function handle() {
		do_action( 'action_scheduler_run_queue', 'Async Request' ); // run a queue in the same way as WP Cron, but declare the Async Request context.

		$sleep_seconds = $this->get_sleep_seconds();

		if ( $sleep_seconds ) {
			sleep( $sleep_seconds );
		}

		$this->maybe_dispatch();
	}
```

3. Chaining requests: After sleeping, it checks if there are more pending actions:

```php
	public function maybe_dispatch() {
		if ( ! $this->allow() ) {
			return;
		}

		$this->dispatch();
		ActionScheduler_QueueRunner::instance()->unhook_dispatch_async_request();
```

4. The check that triggers chaining:

```php
protected function allow() {

		if ( ! has_action( 'action_scheduler_run_queue' ) || ActionScheduler::runner()->has_maximum_concurrent_batches() || ! $this->store->has_pending_actions_due() ) {
			$allow = false;
		} else {
			$allow = true;
		}

		return apply_filters( 'action_scheduler_allow_async_request_runner', $allow );
	}
```

## Why this causes performance issues

- Each async request blocks a PHP-FPM worker for 5+ seconds (sleep + processing)
- With many pending actions, this creates a chain of blocking requests
- On a site with 180k posts, there may be many scheduled actions, causing continuous chaining
- This can exhaust PHP-FPM workers, causing site slowdowns

## Root cause

The async request runner is intended for smaller sites. For large sites (180k+ posts), use WP-CLI or system cron instead.

# Recommended solutions

## 1. Disable Async Request Runner

Add this to the client's theme's functions.php or a custom plugin:

```php
// Disable Action Scheduler async request runner for better performance
add_filter( 'action_scheduler_allow_async_request_runner', '__return_false' );
```

Then set up WP-CLI or system cron to run the queue instead.

## 2. Use WP-CLI instea

Set up a system cron job to run:

```bash
wp action-scheduler action run $( wp action-scheduler action list --status=pending --group=action-scheduler --format=ids )
```

## 3. Temporarily reduce the sleep time

If async requests must be used, reduce the sleep time:

```php
// Reduce sleep time from 5 seconds to 1 second
add_filter( 'action_scheduler_async_request_sleep_seconds', function() {
    return 1; // Reduce from default 5 seconds
});
```

Warning: This may increase MySQL load if there are many chained requests.

## 4. Check duplicated scheduler

Check if there are duplicate Action Scheduler instances (the client mentioned "duplicate schedulers").

```php
// Check if multiple plugins are loading Action Scheduler
add_action('plugins_loaded', function() {
    if (class_exists('ActionScheduler')) {
        error_log('Action Scheduler loaded from: ' . ActionScheduler::plugin_path());
    }
}, 999);

```

This will help identify if multiple plugins are bundling Action Scheduler, which could cause duplicate processing.

# Support ticket

https://wordpress.org/support/topic/sluggishness-with-180k-posts/

# Notes

If we find a good fix for this, it worths evaluating the option to include a settings page related to performance, where the user can disable async requests, or have instructions regarding how to setup the cron job's command.

**Comment by andergmartins** [comment]
This one is particularly important because we are considering add that library to other plugins we have. On that case, we need a clear position of a fix for larger sites.
